File: Lab 12/12-2.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
    
        graph = {i: [] for i in range(numCourses)}
        for course, prereq in prerequisites:
            graph[course].append(prereq)
        
        visited = [0] * numCourses
        
        def dfs(course):
            if visited[course] == 1:  
                return False
            if visited[course] == 2:  
                return True
            
            visited[course] = 1
            for prereq in graph[course]:
                if not dfs(prereq):
                    return False
            
            visited[course] = 2
            return True
        
        for course in range(numCourses):
            if not dfs(course):
                return False
        
        return True

File: Lab 12/12-2.java
import java.util.*;

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] prereq : prerequisites) {
            graph.get(prereq[0]).add(prereq[1]);
        }
        
        int[] visited = new int[numCourses];
        
        for (int i = 0; i < numCourses; i++) {
            if (!dfs(i, visited, graph)) {
                return false; 
            }
        }
        
        return true; 
    }
    
    private boolean dfs(int course, int[] visited, List<List<Integer>> graph) {
        if (visited[course] == 1) {
            return false; 
        }
        if (visited[course] == 2) {
            return true; 
        }
        
        visited[course] = 1;
        for (int prereq : graph.get(course)) {
            if (!dfs(prereq, visited, graph)) {
                return false;
            }
        }
        
        visited[course] = 2;
        return true;
    }
}
